mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp

```diff
@@ -3295,6 +3220,318 @@ LogicalResult OpenMPDialectLLVMIRTranslationInterface::convertOperation(
+      });
+}
+
+static LogicalResult
+convertHostOp(Operation *op, llvm::IRBuilderBase &builder,
+                     LLVM::ModuleTranslation &moduleTranslation) {
+  return convertCommonOperation(op, builder, moduleTranslation);
+}
+
+/// Implementation of the dialect interface that converts operations belonging
+/// to the OpenMP dialect to LLVM IR.
+class OpenMPDialectLLVMIRTranslationInterface
+    : public LLVMTranslationDialectInterface {
+private:
+  mutable ConversionDispatchList dispatchList;
+
+public:
+  using LLVMTranslationDialectInterface::LLVMTranslationDialectInterface;
+
+  /// Translates the given operation to LLVM IR using the provided IR builder
+  /// and saving the state in `moduleTranslation`.
+  LogicalResult
+  convertOperation(Operation *op, llvm::IRBuilderBase &builder,
+                   LLVM::ModuleTranslation &moduleTranslation) const final;
+
+  /// Given an OpenMP MLIR attribute, create the corresponding LLVM-IR, runtime
+  /// calls, or operation amendments
+  LogicalResult
+  amendOperation(Operation *op, ArrayRef<llvm::Instruction *> instructions,
+                 NamedAttribute attribute,
+                 LLVM::ModuleTranslation &moduleTranslation) const final;
+};
+
+// Implementation converting a nest of operations in a single function. This
+// just overrides the parallel and wsloop dispatches but does the normal
+// lowering for now.
+static LogicalResult convertOmpDistributeParallelWsLoop(
+    Operation *op, omp::DistributeOp distribute, omp::ParallelOp parallel,
+    omp::WsLoopOp wsloop, llvm::IRBuilderBase &builder,
+    LLVM::ModuleTranslation &moduleTranslation,
+    ConversionDispatchList &dispatchList) {
+
+  // Convert parallel alternative implementation
+  ConvertFunctionTy convertParallel =
+      [](Operation *op, llvm::IRBuilderBase &builder,
+         LLVM::ModuleTranslation &moduleTranslation) {
+        // Could potentially check that this op is exactly the same parallel
+        // op passed in
+        if (!isa<omp::ParallelOp>(op)) {
+          return std::make_pair(false, failure());
+        }
+        llvm::errs() << "====== Matching Alternative Parallel Lowering =====\n";
+        LogicalResult result =
+            convertCommonOperation(op, builder, moduleTranslation);
+        return std::make_pair(true, result);
+      };
+
+  // Convert wsloop alternative implementation
+  ConvertFunctionTy convertWsLoop =
+      [](Operation *op, llvm::IRBuilderBase &builder,
+         LLVM::ModuleTranslation &moduleTranslation) {
+        if (!isa<omp::WsLoopOp>(op)) {
+          return std::make_pair(false, failure());
+        }
+        llvm::errs() << "====== Matching Alternative WsLoop Lowering =====\n";
+        LogicalResult result =
+            convertCommonOperation(op, builder, moduleTranslation);
+        return std::make_pair(true, result);
+      };
+
+  // Push the new alternative functions
+  dispatchList.pushConversionFunction(convertParallel);
```

Comment:
These functions are called when ``convertOperation`` is called on ``dispatchList``, but that seems to be happening before we get into ``convertOmpDistributeParallelWsLoop``. Am I missing something?

Reply:
Initially this list is empty and only is populateed if there are enclosing op that causes inner ops to be lowered differently and need modified conversion functions.